### PR TITLE
Add mobile crash data model spec

### PR DESCRIPTION
**Pull request by tonzhan2**
Tentative draft for mobile crashes reflected as standalone event logs. Not sure if added in the right place, and a little unfamiliar with formatting so any input is appreciated

### CODE REVIEW

**Comment by AlexanderWert** [comment]
cc @open-telemetry/semconv-mobile-approvers

File: model/logs/mobile-crashes.yaml
Comment (nachoBonafonte) [inline model/logs/mobile-crashes.yaml:40]: Why the content of PLCrashReporter and not an Apple style crash file? which can be generated not only from PLCrashreporter but from others like KSCrash. Apart from supporting other crash handlers it would also support existing tools for symbolication

File: model/logs/mobile-crashes.yaml
Comment (breedx-splk) [inline model/logs/mobile-crashes.yaml:11]: I think the description for `platform` is incorrect here.

File: model/logs/mobile-crashes.yaml
Comment (SMickelsn) [reply model/logs/mobile-crashes.yaml:40]: We do actually use PLCrashReporter with the Apple symbolication style.

File: model/logs/mobile-crashes.yaml
Comment (LikeTheSalad) [inline model/logs/mobile-crashes.yaml:6]: I think the domain should be `device` which is the one targeting mobile apps in the [events spec](https://opentelemetry.io/docs/specs/otel/logs/semantic_conventions/events/).

File: model/logs/mobile-crashes.yaml
Comment (LikeTheSalad) [inline model/logs/mobile-crashes.yaml:38]: I think nested attributes haven't been approved into the OTel spec yet if I'm not mistaken, so if we wanted to use them we should probably add that dependency on this PR.

Another alternative (also not approved) would be supporting structured log bodies that would allow for nested properties. There's a prototype on it in the [Java repo](https://github.com/open-telemetry/opentelemetry-java/pull/5880) in case anyone's interested!

File: model/logs/mobile-crashes.yaml
Comment (breedx-splk) [inline model/logs/mobile-crashes.yaml:17]:
```suggestion
              brief: 'Value if crash happened on an android device'
```

File: model/logs/mobile-crashes.yaml
Comment (breedx-splk) [inline model/logs/mobile-crashes.yaml:20]:
```suggestion
              brief: 'Value if crash happened on an ios device'
```

File: model/logs/mobile-crashes.yaml
Comment (breedx-splk) [inline model/logs/mobile-crashes.yaml:33]: Wait, can't this just be boolean?
```suggestion
        type: boolean
```

File: model/logs/mobile-crashes.yaml
Comment (surbhiia) [inline model/logs/mobile-crashes.yaml:11]:
```suggestion
            This attribute indicates the operating system name of the device on which the crash event occurred.
```

File: model/logs/mobile-crashes.yaml
Comment (surbhiia) [reply model/logs/mobile-crashes.yaml:40]: May be the description can be made generic to incorporate all. Something like - "On iOS, it will contain 'crashReport' (obtained from iOS crash reporters like PLCrashReporter, KSCrash etc.)"

**Review by breedx-splk — APPROVED** [review]

**Review by surbhiia — APPROVED** [review]

File: model/logs/mobile-crashes.yaml
Comment (breedx-splk) [reply model/logs/mobile-crashes.yaml:30]: I'm sure it's slightly contentious, but the _stable_ logs data model does [specify attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/eda62e4baf906ac8225a9ee69fa97ff7f48dbccc/specification/logs/data-model.md#field-attributes) as `map<string,any>`. Contrast this with the [common attributes spec](https://github.com/open-telemetry/opentelemetry-specification/tree/eda62e4baf906ac8225a9ee69fa97ff7f48dbccc/specification/common#attribute) that constrains the values that attributes can have (primitives or homogeneous arrays only). 

The common attributes spec [discusses mapping arbitrary composite data](https://github.com/open-telemetry/opentelemetry-specification/blob/eda62e4baf906ac8225a9ee69fa97ff7f48dbccc/specification/common/attribute-type-mapping.md#composite-values) into attributes, so there's probably a case to be made.

Comment (AlexanderWert) [reply model/logs/mobile-crashes.yaml:30]: Why not making it explicit? I.e. reusing `exception.message`, `exception.stackstrace`, `thread.name` / `thread.id` attributes instead of encoding them into a `data` object?

Comment (scheler) [reply model/logs/mobile-crashes.yaml:30]: @tonzhan2  I suggest listing the contents of `event.data` as top level attributes to begin with, to get feedback. And on the side let's continue to ask how to represent nested attributes in semantic conventions.

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:26]: we should rather reference the `exception.stacktrace` attribute here instead:
https://github.com/open-telemetry/semantic-conventions/blob/main/docs/exceptions/exceptions-spans.md#attributes

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:32]: we should rather reference the existing `thread.*` attributes here, instead:
https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/attributes.md#general-thread-attributes

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:38]: Let's reference `exception.message` here:
https://github.com/open-telemetry/semantic-conventions/blob/main/docs/exceptions/exceptions-spans.md#attributes

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:44]:
```suggestion
      - id: crash.report
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:21]: why not reusing the `ios.state` and `android.state` here instead?
https://github.com/open-telemetry/semantic-conventions/blob/main/docs/mobile/events.md

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:8]: Let's use `os.name` instead:
https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/os.md#operating-system

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:47]: What is PLCrashReporter? Is that iOS or Android specific or generic?
Also, the brief seems to be not complete here? `For ios crashes ...`

**Review by AlexanderWert — CHANGES_REQUESTED** [review]
- Except for `crash.report` all other attributes should be replaced with references to existing (semantically very similar) attributes.
- For new attributes, please add their definition under `https://github.com/open-telemetry/semantic-conventions/tree/main/model/registry` (in the proper namespace file) and then reference them in this file.
- Please add a readme for the mobile-crashes semantic conventions and generate a table with attributes from this file using `make table-generation`

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:8]: @tonzhan2 When you are reusing existing attributes, you need to reference them. (Here is an example on how to do that: https://github.com/open-telemetry/semantic-conventions/blob/39af37d0935f7e9030cddade5b0e16a9c4ea135c/model/trace/http.yaml#L10).

In particular, when you are specifying `id` and `type` you are defining an attribute, but those attributes already exist. So, that would cause some conflicts.

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:15]:
```suggestion
        requirement_level: 
            recommended: If it is a crash on iOS
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:22]:
```suggestion
        requirement_level: 
            recommended: If it is a crash on Android
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:14]:
```suggestion
        requirement_level:
          recommended: If the crash occurred on an iOS device.
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:15]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:16]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:17]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:18]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:24]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:23]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:22]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:21]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:20]:
```suggestion
        requirement_level:
          recommended: If the crash occurred on an Android device.
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:6]:
```suggestion
      `event.domain` is 'device' and `event.name` is 'crash'.
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:8]: I think we can omit `os.name` here because `os.name` will be capture in the resource attributes anyways.

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:26]:
```suggestion
        requirement_level:
          recommended: If the crash occurred on an Android device.
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:29]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:34]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:31]:
```suggestion
        requirement_level:
          recommended: If the crash occurred on an Android device.
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:39]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:36]:
```suggestion
        requirement_level:
          recommended: If the crash occurred on an Android device.
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:44]:
```suggestion
```

Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:44]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:43]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:42]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:41]:
```suggestion
        requirement_level:
          recommended: If the crash occurred on an Android device.
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:40]:
```suggestion
      - ref: exception.message
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:45]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:46]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:48]:
```suggestion
        requirement_level:
          recommended: If the crash occurred on an iOS device.
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:54]: Can you add a real (maybe shortened) example?

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:51]: What is meant here by "These could be broken out as individual elements"?

Do you mean put them into separate attributes?

**Review by AlexanderWert — CHANGES_REQUESTED** [review]

**Comment by AlexanderWert** [comment]
@tonzhan2 I added a bunch of review comments / proposals on how to refactor the yaml file to be consistent with other semantic conventions.

Apart from that, there are two major things left:
1. Can you add a readme file (in `semantic-conventions/docs/mobile`) that would describe the semantic conventions and generate the attributes table in there (see other markdown files for reference).
   - you'll need the `<!-- semconv [SEMCONV_ID_GOES_HERE]--> ... <!-- endsemconv -->` comments
   - and then run `make table-generation`
2. Currently the payload for Android crashes is in `exception.*` and `thread.*` attributes, while for iOS it's a single field where the crash content goes in. It might be fine as the data is conceptually different for those systems, but I'd like to here more opinions from @open-telemetry/semconv-mobile-approvers

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:8]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:9]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:10]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:11]:
```suggestion
```

File: model/logs/mobile-crashes.yaml
Comment (AlexanderWert) [inline model/logs/mobile-crashes.yaml:12]:
```suggestion
```

**Comment by tonzhan2** [comment]
> @tonzhan2 I added a bunch of review comments / proposals on how to refactor the yaml file to be consistent with other semantic conventions.
> 
> Apart from that, there are two major things left:
> 
> 1. Can you add a readme file (in `semantic-conventions/docs/mobile`) that would describe the semantic conventions and generate the attributes table in there (see other markdown files for reference).
>    
>    * you'll need the `<!-- semconv [SEMCONV_ID_GOES_HERE]--> ... <!-- endsemconv -->` comments
>    * and then run `make table-generation`
> 2. Currently the payload for Android crashes is in `exception.*` and `thread.*` attributes, while for iOS it's a single field where the crash content goes in. It might be fine as the data is conceptually different for those systems, but I'd like to here more opinions from @open-telemetry/semconv-mobile-approvers

Will do.
I would defer any questions about the iOS crash payload in the single field, as well as details pertaining to the iOS formatting to my colleague @SMickelsn

File: model/logs/mobile-crashes.yaml
Comment (SMickelsn) [reply model/logs/mobile-crashes.yaml:38]: Images and Threads are arrays containing details of objects linked and stack traces respectively.
The signal elements are broken out separately.   I will add a sample to the previous comment

File: model/logs/mobile-crashes.yaml
Comment (SMickelsn) [reply model/logs/mobile-crashes.yaml:46]: So sorry for the very long delay.

Here is a sample of the data generated, with the threads and images arrays collapsed for space.

  "crashReport" : {
    "cpuCount" : 10,
    "appVersion" : "0.9(5)",
    "parentProcessName" : "launchd_sim",
    "faultAddress" : "6751435752",
    "isNative" : "1",
    "processPath" : "\/Users\/mickelsn\/Library\/Developer\/CoreSimulator\/Devices\/DB54D056-8CE5-4841-87A0-44664A7E19DB\/data\/Containers\/Bundle\/Application\/052B1F60-CD75-4978-8C19-7955B0354D0B\/AppDCloudEveryFeature.app\/AppDCloudEveryFeature",
    "cpuLogicalCount" : 10,
    "parentProcessId" : 54459,
    "osBuild" : "23C71",
    "timestamp" : "2024-01-22 20:58:07 +0000",
    "threads" : [

    ],
    "osVersion" : "17.0.1",
    "signalCode" : "TRAP_BRKPT",
    "appMinorVersionAtCrash" : "5",
    "actualTimestamp" : "2024-01-22 20:58:07 +0000",
    "cpuType" : {
      "subType" : "0",
      "cpuType" : "16777228"
    },
    "osVersionAtCrash" : "17.0.1",
    "hardwareModel" : "arm64",
    "processName" : "AppDCloudEveryFe",
    "appBundleId" : "com.AppDynamics.AppDCloudEveryFeature",
    "images" : [
      
	],
    "signalName" : "SIGTRAP",
    "operatingSystem" : 2,
    "appMajorVersionAtCrash" : "0.9",
    "processId" : 99932
  }

**Comment by joaopgrassi** [comment]
CC @marandaneto

File: model/logs/mobile-crashes.yaml
Comment (marandaneto) [inline model/logs/mobile-crashes.yaml:37]: Would be possible to make the reporter agnostic?
In case I use [KSCrash](https://github.com/kstenerud/KSCrash), I'd like to use the very same data model.
Maybe introducing a new property that tells what's the reporter format, such as `KSCrash`, `PLCrashReporter` or anything else.

File: model/logs/mobile-crashes.yaml
Comment (marandaneto) [inline model/logs/mobile-crashes.yaml:14]: Maybe we should just call it `state` and allow the OS name to be set such as `ios` or `android`, at least is flexible enough in case a new OS appears rather than introducing attributes for every OS.

File: model/logs/mobile-crashes.yaml
Comment (marandaneto) [inline model/logs/mobile-crashes.yaml:11]: `ios` is specifically for iPhone, `macOS` is for MacBook and friends.
Maybe we call it `apple` instead? since the crash reports share the same format on all Apple platforms.

File: model/logs/mobile-crashes.yaml
Comment (marandaneto) [inline model/logs/mobile-crashes.yaml:7]: `event.name` should allow something different than `crash`.
`crash` means the application exited, but this format should allow swallowed errors as well which is technically not a `crash` but rather an error or exception.

Comment (scheler) [reply model/logs/mobile-crashes.yaml:7]: The current thinking is that we will model a separate event type for non-fatal errors/exceptions. There is already an event type called `exception` in the spec but that's more for span events, that is, errors occuring in the context of a span, but we will need to model a similar event for a more generic usecase.

**Pull request closed without merging**
